### DISCUSSION TITLE
Which libsodium version adds XChaCha20Poly1305 to crypto_secretbox ?

### DISCUSSION THREAD

**Discussion by Chewhern** [jedisct1/libsodium#1108] [Q&A] [answered]
When I code C# binding for libsodium not long ago, I realized that in libsodium official documentation(website/github) does not include XChaCha20Poly1305. I would like to know which version was XChaCha20Poly1305 added to the library?

**Comment by jedisct1** [comment, marked as answer]
According to the [ChangeLog](https://github.com/jedisct1/libsodium/blob/64129657a5c67f3bab84562aa8d57dacc685cc75/ChangeLog#L180-L181), it was in version 1.0.12.

**Reply by Chewhern** [reply]
Do you mind if I create a merge pull request(fan made version) to add XChaCha20Poly1305(secretbox) into libsodium doc as I personally think that most bindings don't include the XChaCha20Poly1305(secretbox)?

**Reply by jedisct1** [reply]
Maybe somewhere in the `advanced` section, even though I'm not sure it is very useful.

The only use case for it I know of for it is the DNSCrypt protocol.

Virtually all applications should either use the high-level secretbox for compatibility with NaCl, or AEADs.

secretbox_xchacha20poly1305 has the same practical security and speed as the normal secretbox.

**Reply by Chewhern** [reply]
I know stream cipher alone exist in ``Advanced`` section. However, those are not what I am referring to. In ``secret-key cryptography`` there's ``authenticated encryption`` which uses ``crypto_secretbox_easy()`` or ``crypto_secretbox_detached()``. However, I guess most of the binding developer don't know there's ``crypto_secretbox_xchacha20poly1305_easy()`` or ``crypto_secretbox_xchacha20poly1305_detached``. What I am suggesting adding was the xchacha20poly1305 variant of ``crypto_secretbox_easy`` and ``crypto_secretbox_xchacha20poly1305_detached()`` to this website (https://doc.libsodium.org/) or to this specific part of github repository - (https://github.com/jedisct1/libsodium-doc/tree/master/secret-key_cryptography). In my opinion, I guess it's better to let developer know there's xchacha20poly1305 variant of secretbox and they can decide whether to include it into their binding or ignore.

**Reply by Chewhern** [reply]
Google also encourage people to use ChaCha20 and some open source operating system also started to use ChaCha20 as a way to generate cryptography random number. Computerphile(not sure if it can be used as a reference here) also publishes a video regarding ChaCha20 as an alternative to AES

**Reply by jedisct1** [reply]
There's nothing wrong with `crypto_secretbox_easy()` and `crypto_secretbox_detached()`.

**Reply by Chewhern** [reply]
I don't mean replacing ... I mean adding another version of secretbox which uses xchacha20poly1305 so there're 2 authenticated encryption instead of only a single authenticated encryption which uses xsalsa20poly1305. Do you think I should or shouldn't make a pull request?

**Reply by jedisct1** [reply]
Yes, the function is probably going to be removed. The only reason it's there is DNSCrypt, really.

**Reply by Chewhern** [reply]
Okay.... So does it mean I should remove it from my binding as well? The xchacha20poly1305 variant of secretbox

**Reply by jedisct1** [reply]
You can keep them, but this is something to use only if you are explicitly looking for it.
